### PR TITLE
IS-316: Refactor score deployment

### PR DESCRIPTION
**Pull request by goldworm-icon**
Add test codes for transaction result inconsistency by undefined score member variable
Remove builtin specific functions from IconScoreDeployEngine
Add ZERO_TX_HASH to icon_constant.py
Replace None with ZERO_TX_HASH for IconScoreDeployInfo and IconScoreInfo
Replace IconScoreInfo.create_score() with IconScoreInfo.get_score(revision)
Remove files and directories in score_deploy_path before deploing a builtin score
Replace IconScoreDeployer.remove_existing_score() with utils.remove_path()
Change IconScoreDeployer.deploy() paramsters: (Address, content, tx_hash) -> (path, content)
Remove score_package_validate codes from IconScoreMapper and IconScoreClassLoader
Move score_package_validate related condes to IconScoreContextUtil
Move score_root_path to IconScoreContext
Implement is_score_active() and get_owner() on IconScoreContextUtil
Add test codes for transaction result inconsistency by undefined score member variable
Update IconScoreDeployEngine._initialize_score()
Update SCORE loading related codes in IconScoreContextUtil
Remove SCORE loading from IconScoreMapper
Pass MainNet and TestNet DB sync test

### CODE REVIEW

File: iconservice/builtin_scores/governance/governance.py [hunk 1]
```diff
@@ -52,6 +52,7 @@
 
 CONTEXT_TYPE_INVOKE = 'invoke'
 CONTEXT_TYPE_QUERY = 'query'
+ZERO_TX_HASH = bytes(32)
```
Comment (leeheonseung) [inline iconservice/builtin_scores/governance/governance.py]: have to add this code from governance SCORE to https://github.com/icon-project/governance after merge to develop

Comment (goldworm-icon) [reply iconservice/builtin_scores/governance/governance.py]: I found a better way to resolve this problem without modifying the existing governance SCORE code.

IconScoreDeployStorage.get_tx_hashes_by_score_address() will convert ZERO_TX_HASH to None instead.
So, we don't need to modify the governance SCORE code anymore.

File: iconservice/builtin_scores/governance/governance.py [hunk 2]
```diff
@@ -196,6 +196,7 @@ def getScoreStatus(self, address: Address) -> dict:
         current_tx_hash, next_tx_hash = self.get_tx_hashes_by_score_address(address)
         if current_tx_hash is None and next_tx_hash is None:
             self.revert('SCORE not found')
+
```
Comment (leeheonseung) [inline iconservice/builtin_scores/governance/governance.py]: should revert

File: iconservice/deploy/__init__.py [hunk 3]
```diff
@@ -24,4 +24,3 @@ class DeployType(IntEnum):
 class DeployState(IntEnum):
     INACTIVE = 0
     ACTIVE = 1
-
```
Comment (leeheonseung) [inline iconservice/deploy/__init__.py]: should revert

File: iconservice/deploy/icon_score_deployer.py [hunk 4]
```diff
@@ -13,49 +13,46 @@
...
 import io
 import os
 import shutil
 import zipfile
 
-from ..base.address import Address
 from ..base.exception import ScoreInstallExtractException, ScoreInstallException
 
 
 class IconScoreDeployer(object):
```
Comment (leeheonseung) [inline iconservice/deploy/icon_score_deployer.py]: should modify name "deploy" -> "extract" it is not important

File: iconservice/icon_constant.py [hunk 5]
```diff
@@ -52,9 +52,12 @@
 
 BUILTIN_SCORE_ADDRESS_MAPPER = {'governance': "cx0000000000000000000000000000000000000001"}
 
+ZERO_TX_HASH = bytes(32)
+
 REVISION_2 = 2
-REVISION_3 = 3
+REVISION_3 = 3  # 1.1.2.x
 REVISION_4 = 4
+LATEST_REVISION = REVISION_3
```
Comment (leeheonseung) [inline iconservice/icon_constant.py]: remove REVISION_4

File: iconservice/icon_service_engine.py [hunk 6]
```diff
@@ -662,10 +653,12 @@ def validate_transaction(self, request: dict) -> None:
 
         self._icon_pre_validator.execute(context, params, step_price, minimum_step)
 
-        self._validate_score_blacklist(context, params)
+        IconScoreContextUtil.validate_score_blacklist(context, to)
         if IconScoreContextUtil.is_service_flag_on(context, IconServiceFlag.DEPLOYER_WHITE_LIST):
             self._validate_deployer_whitelist(context, params)
```
Comment (leeheonseung) [inline iconservice/icon_service_engine.py]: move function to IconScoreContextUtil
"_validate_deployer_whitelist"

remove function _validate_score_blacklist in iconserviceengine

File: iconservice/iconscore/icon_pre_validator.py [hunk 7]
```diff
@@ -259,6 +261,14 @@ def _check_balance(self, context: 'IconScoreContext', from_: 'Address', value: i
     def _is_inactive_score(self, address: 'Address') -> bool:
         is_contract = address.is_contract
         is_zero_score_address = address == ZERO_SCORE_ADDRESS
-        is_score_active = self._deploy_storage.is_score_active(None, address)
+        is_score_active = self._is_score_active(address)
         _is_inactive_score = is_contract and not is_zero_score_address and not is_score_active
         return _is_inactive_score
+
+    def _is_score_active(self, address: 'Address') -> bool:
```
Comment (leeheonseung) [inline iconservice/iconscore/icon_pre_validator.py]: we don't need to use this function.
we have to determine remove it or not

File: iconservice/iconscore/icx.py [hunk 8]
```diff
@@ -74,7 +74,9 @@ def get_balance(self, address: 'Address') -> int:
     # noinspection PyBroadException
     def _is_icx_send_defective(self) -> bool:
         try:
-            governance_score = IconScoreContextUtil.get_icon_score(self._context, GOVERNANCE_SCORE_ADDRESS)
+            governance_score = IconScoreContextUtil.get_builtin_score(
+                self._context, GOVERNANCE_SCORE_ADDRESS)
+
             if governance_score is not None:
```
Comment (leeheonseung) [inline iconservice/iconscore/icx.py]: already check it.
remove duplicate logic

File: iconservice/iconscore/icon_score_context_util.py [hunk 9]
```diff
@@ -34,47 +41,130 @@ class IconScoreContextUtil(object):
...
-        score = None
+        """Create a SCORE instance from SCORE class and return it
+
+        :param context:
+        :param address:
+        :return:
+        """
+        score_info: 'IconScoreInfo' = None
 
         if context.type == IconScoreContextType.INVOKE:
```
Comment (leeheonseung) [inline iconservice/iconscore/icon_score_context_util.py]: remove lagacy logic

File: iconservice/iconscore/icon_score_class_loader.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+        :param score_deploy_path:
+        :return:
+        """
+        pkg_json_path = os.path.join(score_deploy_path, IconScoreClassLoader._PACKAGE_JSON_FILE)
+        with open(pkg_json_path, 'r') as f:
+            return json.load(f)
+
+    @staticmethod
+    def run(score_address: 'Address', tx_hash: bytes, score_root_path: str) -> type:
```
Comment (leeheonseung) [inline iconservice/iconscore/icon_score_class_loader.py]: replace score_address, tx_hash to path

File: iconservice/iconscore/icon_score_class_loader.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    def init(score_root_path: str):
+        if score_root_path not in sys.path:
+            sys.path.append(score_root_path)
+
+    @staticmethod
+    def exit(score_root_path: str):
+        sys.path.remove(score_root_path)
+
+    @staticmethod
+    def _load_package_json(score_deploy_path: str) -> dict:
```
Comment (leeheonseung) [inline iconservice/iconscore/icon_score_class_loader.py]: remove function

File: iconservice/deploy/icon_score_deploy_engine.py [hunk 12]
```diff
@@ -83,72 +81,78 @@ def invoke(self,
...
-        return not is_audit_enabled or all((is_system_score, is_owner))
+        # FiXME: SCORE owner check should be done before calling self._is_audit_needed().
+        is_owner: bool = context.tx.origin == IconScoreContextUtil.get_owner(context, score_address)
+        is_audit_enabled: bool = IconScoreContextUtil.is_service_flag_on(context, IconServiceFlag.AUDIT)
+
+        return is_audit_enabled and not (is_system_score and is_owner)
 
     def deploy(self, context: 'IconScoreContext', tx_hash: bytes) -> None:
         """
-        included audit deploy
```
Comment (leeheonseung) [inline iconservice/deploy/icon_score_deploy_engine.py]: set 0.write deployinfo and deployTxparams to statedb

File: iconservice/deploy/icon_score_deploy_engine.py [hunk 13]
```diff
@@ -83,72 +81,78 @@ def invoke(self,
...
         :param tx_params: use deploy_data from IconScoreDeployTxParams info
         :return:
         """
 
-        data = tx_params.deploy_data
-        content_type = data.get('contentType')
+        data: dict = tx_params.deploy_data
+        content_type: str = data.get('contentType')
 
         if content_type == 'application/tbears':
```
Comment (leeheonseung) [inline iconservice/deploy/icon_score_deploy_engine.py]: should divide function

File: iconservice/deploy/icon_score_deploy_engine.py [hunk 14]
```diff
@@ -244,121 +176,132 @@ def _on_deploy(self,
...
+            write_score_to_score_deploy_path: callable =\
+                self._write_score_to_score_deploy_path
+        write_score_to_score_deploy_path(context, score_address, next_tx_hash, content)
 
         backup_msg = context.msg
         backup_tx = context.tx
+        new_score_mapper: 'IconScoreMapper' = context.new_icon_score_mapper
 
         try:
             if IconScoreContextUtil.is_service_flag_on(context, IconServiceFlag.SCORE_PACKAGE_VALIDATOR):
```
Comment (leeheonseung) [inline iconservice/deploy/icon_score_deploy_engine.py]: make function one try_score_package_validate
insert this logic to try_score_package_validate

File: iconservice/deploy/icon_score_deploy_engine.py [hunk 15]
```diff
@@ -244,121 +176,132 @@ def _on_deploy(self,
...
-            deploy_type = tx_params.deploy_type
-            if deploy_type == DeployType.INSTALL:
-                on_deploy = score.on_install
-            elif deploy_type == DeployType.UPDATE:
-                on_deploy = score.on_update
-            else:
-                on_deploy = None
+                IconScoreContextUtil.validate_score_package(context, score_address, next_tx_hash)
+
+            score_info: 'IconScoreInfo' =\
```
Comment (leeheonseung) [inline iconservice/deploy/icon_score_deploy_engine.py]: consider revision check first?

File: iconservice/deploy/__init__.py (hunk 3 above)
Comment (goldworm-icon) [reply iconservice/deploy/__init__.py]: An empty line was added to the bottom to comply with pep8.

File: iconservice/icon_constant.py (hunk 5 above)
Comment (leeheonseung) [reply iconservice/icon_constant.py]: we are going to make new task for fix at one time.
so you don't need to fix now

File: iconservice/iconscore/icon_pre_validator.py (hunk 7 above)
Comment (goldworm-icon) [reply iconservice/iconscore/icon_pre_validator.py]: IconPreValidator class refactoring will go on in another branch later.

File: iconservice/iconscore/icx.py (hunk 8 above)
Comment (goldworm-icon) [reply iconservice/iconscore/icx.py]: Remove "if governance_score is not None"

File: iconservice/iconscore/icon_score_class_loader.py (hunk 10 above)
Comment (goldworm-icon) [reply iconservice/iconscore/icon_score_class_loader.py]: Keep this code unchanged. More detail review is needed.

File: iconservice/iconscore/icon_score_class_loader.py (hunk 11 above)
Comment (goldworm-icon) [reply iconservice/iconscore/icon_score_class_loader.py]: Keep this code unchanged.

File: iconservice/deploy/icon_score_deploy_engine.py (hunk 15 above)
Comment (goldworm-icon) [reply iconservice/deploy/icon_score_deploy_engine.py]: If there is no critical bug, keep it unchanged.

File: iconservice/deploy/icon_score_deploy_engine.py (hunk 14 above)
Comment (goldworm-icon) [reply iconservice/deploy/icon_score_deploy_engine.py]: Done

**Pull request merged**